### ISSUE TITLE
A pair of name collision bugs

### ISSUE THREAD

**Issue by zenhack** [zenhack/haskell-capnp#65]
This minimal schema exhibits the problem:

```capnp
@0xa0b17c61e808176b;

struct S {
  union {
    a :group {
      c @2 :Void;
      union {
        d @0 :Void;
        e @1 :Void;
      }
    }
    b @3 :Void;
  }
}
```

You end up with two data constructors called `S'a`, one for the outermost union's `a` variant, and one for the group it contains.

When a group doesn't itself have an anonymous union, we dodge this by collapsing its fields into the parent. We need to work out what to do if it does.

One approach would be to just collapse it anyway. But this brings up another bug we have: if two groups have fields with the same name but different types, the resulting code will not compile, since even with DuplicateRecordFields, if two variants of the same datatype have the same record label, they must have the same type. The result is code that doesn't build. Example of the latter issue:

```capnp
struct CommonFieldNames {
  union {
    a :group {
      c @0 :Void;
    }
    b :group {
      c @1 :Int32;
    }
  }
}
```

**Comment by zenhack** [comment]
The alternative would be to never collapse fields, which solves the second bug, and come up with some other convention for avoiding name collisions.

**Comment by zenhack** [comment]
I went with said alternative. Instead, we append a single-quote to the data constructors for groups, so we have e.g.

```haskell
data Some'group = Some'group' { ... }
```

This causes a lot of build time breakage that is very mechanical to fix; bd49a123723f41119c7f13cc827a5ed2925e94d5 does this for the code generator and test suite, and probably provides a good guide for anyone else trying to upgrade.
